scanpy/external/tl/_annotator.py

```diff
@@ -0,0 +1,146 @@
+"""\
+Annotates gene expression (cell data) with cell types.
+"""
+import warnings
+
+from anndata import AnnData
+import pandas as pd
+
+
+def annotator(
+    adata: AnnData,
+    markers: pd.DataFrame,
+    num_genes: int = None,
+    return_nonzero_annotations: bool = True,
+    p_threshold: float = 0.05,
+    p_value_fun: str = "binom",
+    z_threshold: float = 1.0,
+    scoring: str = "exp_ratio",
+    normalize: bool = False,
+):
+    """\
+    Annotator marks the data with cell type annotations based on marker genes.
+
+    Over-expressed genes are selected with the Mann-Whitney U tests and cell
+    types are assigned with the hypergeometric test. This function first selects
+    genes from gene expression data with the Mann-Whitney U test, then annotate
+    them with the hypergeometric test, and finally filter out cell types that
+    have zero scores for all cells. The results are scores that tell how
+    probable is each cell type for each cell.
+
+    Parameters
+    ----------
+    adata
+        Tabular data with gene expressions.
+    markers
+        The data-frame with marker genes and cell types. Data-frame has two
+        columns **Gene** and **Cell Type** first holds gene names or ID and
+        second cell type for this gene. Gene names must be written in the same
+        format than genes in `adata`.
+    num_genes
+        The number of genes that the organism has.
+    return_nonzero_annotations
+        If true return scores only for cell types that have no zero scores.
+    p_threshold
+        A threshold for accepting the annotations. Annotations that have FDR
+        value bellow this threshold are used.
+    p_value_fun
+        A function that calculates a p-value. It can be either
+        `binom` that uses binom.sf or
+        `hypergeom` that uses hypergeom.sf.
+    z_threshold
+        The threshold for selecting the gene from gene expression data.
+        For each cell the attributes with z-value above this value are selected.
+    scoring
+        Scoring method for cell type scores. Available scores are:
+
+        exp_ratio
+            Proportion of genes typical for a cell type expressed in the cell
+        sum_of_expressed_markers
+            Sum of expressions of genes typical for a cell type
+        log_fdr
+            Negative of the logarithm of an false discovery rate (FDR) value
+        log_p_value
+            Negative of the logarithm of a p-value
+    normalize : bool, optional (default = False)
+        If this parameter is True data will be normalized during the
+        a process with a log CPM normalization.
+        That method works correctly data needs to be normalized.
+        Set this `normalize` on True if your data are not normalized already.
+
+    Returns
+    -------
+    pd.DataFrame
```

Comment:
That’s not right, it returns an `AnnData` object.

Also please orient yourself at the way other scanpy functions work: They have an `inplace` or `copy` argument that controls if a copy of the original `adata` object is made or `obs`/`var` is updated in the original.

This function filters currently out cells, but it would be better to just mark them as having no marker cell type found.

Reply:
@flying-sheep, I find this comment is a little confusing. `inplace=False` will often return some other dtype. `copy` determines whether the anndata object is copied, and that copy modified.

Reply:
Let’s be perfectly clear: `inplace` and `copy` are both used, sometimes `copy` is being deprecated in favor of `inplace`.

`copy` means “if copy=True, return a copy of the AnnData object, with var/obs modified. if copy=False, return `None` and modify the AnnData object’s var/obs”

`inplace` means “if inplace=True, modify the AnnData object, if inplace=False, return an object with computation results”. There’s two variants for when it’s True: a) the AnnData object’s obs/var get new columns or b) the AnnData object is filtered.

I updated my comment above

Reply:
I added an in-place argument